src/box/sql/mem.c

```diff
@@ -2043,14 +2155,86 @@ mem_sub(const struct Mem *left, const struct Mem *right, struct Mem *result)
 	return 0;
 }
 
+/**
+ * Subtract the second MEM from the first MEM and write the result to the third
+ * MEM. The first MEM should be of DATETIME type and the second MEMs should be
+ * of INTERVAL or DATETIME type. The result is of INTERVAL type if the second
+ * MEM is of DATETIME type and the result if of DATETIME type if the second
+ * argument if of INTERVAL type.
+ */
+static int
+mem_sub_dt(const struct Mem *left, const struct Mem *right, struct Mem *result)
+{
+	assert(mem_is_datetime(left) && !mem_is_metatype(left));
+	if (!(mem_is_interval(right) || mem_is_datetime(right)) ||
+	    mem_is_metatype(right)) {
+		diag_set(ClientError, ER_SQL_TYPE_MISMATCH, mem_str(right),
+			 "datetime or interval");
+		return -1;
+	}
+	if (mem_is_datetime(right)) {
+		struct interval res;
+		if (datetime_datetime_sub(&res, &left->u.dt, &right->u.dt) != 0)
+			return -1;
+		mem_set_interval(result, &res);
+		return 0;
+	}
+	struct datetime res = left->u.dt;
+	if (datetime_increment_by(&res, -1, &right->u.itv) != 0)
+		return -1;
+	mem_set_datetime(result, &res);
+	return 0;
+}
+
+/**
+ * Subtract the second MEM from the first MEM and write the result to the third
+ * MEM. The first and the second MEMs should be of INTERVAL types. The result is
+ * of INTERVAL type.
+ */
+static int
+mem_sub_itv(const struct Mem *left, const struct Mem *right, struct Mem *result)
+{
+	assert(mem_is_interval(left) && !mem_is_metatype(left));
+	if (!mem_is_interval(right) || mem_is_metatype(right)) {
+		diag_set(ClientError, ER_SQL_TYPE_MISMATCH, mem_str(right),
+			 "interval");
+		return -1;
+	}
+	struct interval res = left->u.itv;
+	if (interval_interval_sub(&res, &right->u.itv) != 0)
+		return -1;
+	mem_set_interval(result, &res);
+	return 0;
```

Comment:
I suggest to update the subtractions too, like the additions. Not to create a temporary interval nor datetime on the stack.

Reply:
Fixed, thank you. Sorry, I forgot that we have addition and subtraction.